使用源码编译Android的PaddlePaddle库报错

使用源码编译Android的PaddlePaddle库报错，报错如下：
```
[ 94%] Linking CXX shared library libpaddle_capi_shared.so
/mnt/android/linux/arm_standalone_toolchain/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: missing group end
/mnt/android/linux/arm_standalone_toolchain/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: use the --help option for usage information
clang38++: error: linker command failed with exit code 1 (use -v to see invocation)
paddle/capi/CMakeFiles/paddle_capi_shared.dir/build.make:240: recipe for target 'paddle/capi/libpaddle_capi_shared.so' failed
make[2]: *** [paddle/capi/libpaddle_capi_shared.so] Error 1
CMakeFiles/Makefile2:1389: recipe for target 'paddle/capi/CMakeFiles/paddle_capi_shared.dir/all' failed
make[1]: *** [paddle/capi/CMakeFiles/paddle_capi_shared.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
```

@Xreki 
这个错误我是今天早上才出现的，之前也没有成功过，但报错是另一种，如下：
```
[ 17%] Completed 'extern_protobuf'
[ 17%] Built target extern_protobuf
error: RPC failed; curl 18 transfer closed with outstanding read data remaining
fatal: The remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed
Cloning into 'protobuf_host'...
error: RPC failed; curl 18 transfer closed with outstanding read data remaining
fatal: The remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed
-- Had to git clone more than once:
          3 times.
CMake Error at /paddle/build_android/third_party/eigen3/tmp/extern_eigen3-gitclone.cmake:40 (message):
  Failed to clone repository: 'https://github.com/RLovelett/eigen.git'


CMakeFiles/extern_eigen3.dir/build.make:90: recipe for target 'third_party/eigen3/src/extern_eigen3-stamp/extern_eigen3-download' failed
make[2]: *** [third_party/eigen3/src/extern_eigen3-stamp/extern_eigen3-download] Error 1
CMakeFiles/Makefile2:284: recipe for target 'CMakeFiles/extern_eigen3.dir/all' failed
make[1]: *** [CMakeFiles/extern_eigen3.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
error: RPC failed; curl 18 transfer closed with outstanding read data remaining
fatal: The remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed
-- Had to git clone more than once:
          3 times.
CMake Error at /paddle/build_android/third_party/protobuf_host/tmp/protobuf_host-gitclone.cmake:40 (message):
  Failed to clone repository: 'https://github.com/qingqing01/protobuf.git'


CMakeFiles/protobuf_host.dir/build.make:89: recipe for target 'third_party/protobuf_host/src/protobuf_host-stamp/protobuf_host-download' failed
make[2]: *** [third_party/protobuf_host/src/protobuf_host-stamp/protobuf_host-download] Error 1
CMakeFiles/Makefile2:390: recipe for target 'CMakeFiles/protobuf_host.dir/all' failed
make[1]: *** [CMakeFiles/protobuf_host.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -166,11 +166,11 @@ include_directories("${CMAKE_CURRENT_BINARY_DIR}/proto")
 include_directories("${CMAKE_CURRENT_BINARY_DIR}/go/pserver/client/c")
 
 set(EXTERNAL_LIBS
-    ${GFLAGS_LIBRARIES}
-    ${GLOG_LIBRARIES}
-    ${CBLAS_LIBRARIES}
-    ${PROTOBUF_LIBRARY}
-    ${ZLIB_LIBRARIES}
+    gflags
+    glog
+    cblas
+    protobuf
+    zlib
     ${PYTHON_LIBRARIES}
 )
 

--- a/cmake/external/zlib.cmake
+++ b/cmake/external/zlib.cmake
@@ -28,7 +28,7 @@ ENDIF(WIN32)
 INCLUDE_DIRECTORIES(${ZLIB_INCLUDE_DIR})
 
 ExternalProject_Add(
-    zlib
+    extern_zlib
     ${EXTERNAL_PROJECT_LOG_ARGS}
     GIT_REPOSITORY  "https://github.com/madler/zlib.git"
     GIT_TAG         "v1.2.8"
@@ -49,9 +49,11 @@ ExternalProject_Add(
                      -DCMAKE_BUILD_TYPE:STRING=${THIRD_PARTY_BUILD_TYPE}
 )
 
+ADD_LIBRARY(zlib STATIC IMPORTED GLOBAL)
+SET_PROPERTY(TARGET zlib PROPERTY IMPORTED_LOCATION ${ZLIB_LIBRARIES})
+ADD_DEPENDENCIES(zlib extern_zlib)
+
 LIST(APPEND external_project_dependencies zlib)
-ADD_LIBRARY(zlib_target STATIC IMPORTED GLOBAL)
-SET_PROPERTY(TARGET zlib_target PROPERTY IMPORTED_LOCATION ${ZLIB_LIBRARIES})
 
 IF(WITH_C_API)
   INSTALL(DIRECTORY ${ZLIB_INCLUDE_DIR} DESTINATION third_party/zlib)

--- a/cmake/generic.cmake
+++ b/cmake/generic.cmake
@@ -104,7 +104,9 @@ function(merge_static_libs TARGET_NAME)
   foreach(lib ${libs})
     list(APPEND libs_deps ${${lib}_LIB_DEPENDS})
   endforeach()
-  list(REMOVE_DUPLICATES libs_deps)
+  if(libs_deps)
+    list(REMOVE_DUPLICATES libs_deps)
+  endif()
 
   # To produce a library we need at least one source file.
   # It is created by add_custom_command below and will helps
@@ -191,10 +193,13 @@ function(cc_library TARGET_NAME)
         list(REMOVE_ITEM cc_library_DEPS warpctc)
         add_dependencies(${TARGET_NAME} warpctc)
       endif()
-      # Support linking flags: --whole-archive (Linux) / -force_load (MacOS)
-      target_circle_link_libraries(${TARGET_NAME} ${cc_library_DEPS})
       if("${cc_library_DEPS}" MATCHES "ARCHIVE_START")
+        # Support linking flags: --whole-archive (Linux) / -force_load (MacOS).
+        # WARNING: Please don't use ARCHIVE_START&ARCHIVE_END if TARGET_NAME will be linked by other libraries.
+        target_circle_link_libraries(${TARGET_NAME} ${cc_library_DEPS})
         list(REMOVE_ITEM cc_library_DEPS ARCHIVE_START ARCHIVE_END)
+      else()
+        target_link_libraries(${TARGET_NAME} ${cc_library_DEPS})
       endif()
       add_dependencies(${TARGET_NAME} ${cc_library_DEPS})
     endif()

